ISPN-14329 Availability of caches should be prevented until a cluster…

jabolina opened the pull request:
… is complete after "shutdown cluster"

https://issues.redhat.com/browse/ISPN-14329
https://issues.redhat.com/browse/ISPN-14414

After the cluster shutdown, the nodes must wait for the complete recovery before accepting commands. We keep the cache's registry as `INITIALIZING` until all nodes join back and the restoration of a stable topology. We do this if the cache has a non-shared store without purging on start.

We log and return an exception with the current members. We can't say the missing member's address for sure, as we only have the persistent UUID.

With the test we added, the log message is:

```text
11:19:12,840 WARN  (jgroups-7,ThreeNodeGlobalStatePartialRestartTest-NodeF:[]) [o.i.t.ClusterCacheStatus] ISPN000681: Recovering cache testCache missing members, currently have [ThreeNodeGlobalStatePartialRestartTest-NodeF, ThreeNodeGlobalStatePartialRestartTest-NodeG] of a total of 3
```

We still need a way to send commands and force the cache to start, even if that means data loss. And handle the inconsistent behavior from reports. That's follow-up work.

But I also wanted to gather some reviews/comments on the approach here (and see how CI behaves). There is also the option to block the cache creation until the cluster is back or receives a manual intervention.

wfink commented:
Blocking the clients is now what I would expect.

For a better user experience it would be nice to have the WARN message ISPN000681 not only at the coordinator but on each starting node.

Each client access will add an ERROR ISPN005003 with full stack trace, as this is expected I would lower it to WARN and remove the stacktrace, this would prevent from a fast growing logfile at startup if there is a huge number of clients ...

Access with the console force a ISPN012005, is it possible to return the state `INITIALIZING` to display the cache state for the console, and lower the message to WARN/INFO without stacktrace for the same reasons

wfink commented:
If the <file-store purge="true"> is set the cache is empty after 'shutdown cluster' and started nodes are all back.
If set to false we have the previous behavior, stale entries are added back if one node goes down and up again.

jabolina commented:
Applied the suggestions to also log the warning message on the client after the join. Another caveat here is that a node is only aware of the node _when_ it joins, that is, the list of members does not update. This can lead to the exception message can be misleading.

I also included a small change in the REST API to handle the initializing case. This way the cache is listed as DEGRADED and the operations (I test get/put) return the exception message of missing members.

wfink commented:
The WARN message ISPN000681 is now shown for all members during start :+1:

wburns commented on core/src/main/java/org/infinispan/factories/AbstractComponentRegistry.java:
```diff
@@ -278,6 +298,10 @@ public void start() {
 
    protected abstract void postStart();
 
+   protected CompletionStage<Void> await() {
```
A more descriptive name would be better here. I thought this was similar to lock await method or something or waiting for something to finish, not just returning a value.

wburns commented on core/src/main/java/org/infinispan/factories/AbstractComponentRegistry.java:
```diff
@@ -248,28 +250,46 @@ public void start() {
-         } catch (Throwable t1) {
-            t.addSuppressed(t1);
-         }
+   private synchronized void putRunning() {
```
```suggestion
   private synchronized void updateStatusRunning() {
```

wburns commented on core/src/main/java/org/infinispan/factories/ComponentRegistry.java:
```diff
@@ -217,6 +221,29 @@ public void start() {
 
+   @Override
+   protected CompletionStage<Void> await() {
+      if (configuration.simpleCache() || icr.running().isInternalCache(cacheName)) return super.await();
```
```suggestion
      if (configuration.simpleCache() || icr.running().isInternalCache(cacheName)) return super.await();
```
You can use this method from CacheMode instead. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/configuration/cache/CacheMode.java#L109

wburns commented on core/src/main/java/org/infinispan/factories/ComponentRegistry.java:
```diff
@@ -217,6 +221,29 @@ public void start() {
+   }
+
+   private boolean waitStableTopology() {
+      // Cache with non-shared stores without purgeOnStart set must wait until a stable topology is installed.
```
Why is this only affecting only stores that aren't shared with persistent state? We should be waiting for the stable topology under any configuration. Otherwise we could have a sub cluster form that doesn't have all the values yet and when the remaining member joins it will not be able to bring along its values.

wburns commented on core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java:
```diff
@@ -813,5 +852,22 @@ CacheTopology getStableTopology() {
+      knownMembers = Immutables.immutableSetConvert(joiners);
+   }
+
+   CompletionStage<Void> subscribe() {
```
Pick a better name here, subscribe means something else for reactive processing.

wburns commented on core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java:
```diff
@@ -813,5 +852,22 @@ CacheTopology getStableTopology() {
+      return Immutables.immutableListConvert(knownMembers);
+   }
+
+   void setCurrentMembers(List<Address> joiners) {
```
They aren't really joiners right?
```suggestion
   void setCurrentMembers(List<Address> members) {
```

wburns commented on core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java:
```diff
@@ -813,5 +852,22 @@ CacheTopology getStableTopology() {
+   }
+
+   List<Address> knownMembers() {
+      return Immutables.immutableListConvert(knownMembers);
```
This seems extraneous, we should just store the immutable List in the field. Also I am not sold do we really need the Set to remove duplicates or can we trust that the invokers of `setCurrentMembers` will send us a list without duplicates? I would rather not have this extra Set used here if possible.

wburns commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -328,7 +328,7 @@ private CacheInfo getCacheInfo(RestRequest request, EmbeddedCacheManager cacheMa
          cacheInfo.status = "IGNORED";
       } else {
-         if(cacheHealth != HealthStatus.FAILED) {
+         if(cacheHealth != HealthStatus.FAILED && cacheHealth != HealthStatus.DEGRADED) {
```
This looks out of place. Is this a different bug?

jabolina replied:
I ended up adding a new status. The problem here is that the cache was not running, which throws an exception.

jabolina commented on core/src/main/java/org/infinispan/manager/DefaultCacheManager.java:
```diff
@@ -995,7 +995,9 @@ public boolean isRunning(String cacheName) {
       CompletableFuture<Cache<?, ?>> cacheFuture = caches.get(cacheName);
       boolean started = cacheFuture != null && cacheFuture.isDone() && !cacheFuture.isCompletedExceptionally();
-      return started && cacheFuture.join().getStatus() == ComponentStatus.RUNNING;
+      if (!started) return false;
+      ComponentStatus cs = cacheFuture.join().getStatus();
+      return cs == ComponentStatus.RUNNING || cs == ComponentStatus.INITIALIZING;
```
I'll revert this and open another JIRA to follow-up. The problem here is that when getting the cache for the REST/CLI API it verifies if the cache is running before applying an operation. Since in our case the cache is still initializing, causes an exception, returning `Unexpected error retrieving data. "ISPN012010: Cache with name 'manual' not found amongst the configured caches"` to the client, which is odd, since we just saw the cache in a list.

jabolina replied:
Created https://issues.redhat.com/browse/ISPN-14414 to handle this.

jabolina commented:
I applied all the suggestions/changes now. An additional point of attention is that I added another health status to identify the recovering stage, meaning the console needs to be updated to reflect this, too.

I added another exception instead of using the `AvailabilityException` so it was easier to log a warning.

wfink commented on core/src/main/java/org/infinispan/topology/MissingMembersException.java:
```diff
@@ -0,0 +1,9 @@
+
+import org.infinispan.commons.CacheException;
+
+public class MissingMembersException extends CacheException {
```
Would be nice to have  a JavaDoc with more details when this is thrown

tristantarrant commented:
LGTM

jabolina commented:
I'll squash the changes.

wburns reviewed (commented):
Looks good to me. However, are we not missing some way for an administrator to force the cache to RUNNING state if they don't have all the members?

pruivo commented on core/src/test/java/org/infinispan/globalstate/ThreeNodeGlobalStatePartialRestartTest.java:
```diff
@@ -0,0 +1,174 @@
+   @Override
+   public Object[] factory() {
+      return new Object[] {
+            new ThreeNodeGlobalStatePartialRestartTest().withCacheMode(CacheMode.DIST_SYNC).withPartitionHandling(PartitionHandling.ALLOW_READ_WRITES)
```
any plans to add other cache modes and partitions handling?

pruivo reviewed (commented):
LGTM

jabolina commented:
Applied the suggestions, thanks.

jabolina commented:
> Looks good to me. However, are we not missing some way for an administrator to force the cache to RUNNING state if they don't have all the members?

That's a next https://issues.redhat.com/browse/ISPN-14418

jabolina commented:
I am going to add a couple of things:

* The coordinator will have an up-to-date list of members when throwing an exception. Other members will have a possibly outdated list (as of right now);
* If the cluster restores after the shutdown, we **do not** purge the store, even if `purge=true`.

wfink commented:
Coordinator now sends the correct members during initializing.
purge is somehow not happening, but if I add 10 test entries there are 2 missed after 'shutdown cluster' initialization if all nodes are back

jabolina commented:
@wfink I've integrated the keyset fixes to the branch, let me know if it is working as expected.

Also, for ignoring purge after the start, I've updated only the SIFS and SingleFile stores, am I missing something here?

wburns commented on core/src/main/java/org/infinispan/factories/ComponentRegistry.java:
```diff
@@ -217,6 +226,37 @@ public void start() {
+    * Caches that do not need state transfer or are private do not need to delay the start.
+    */
+   @Override
+   protected CompletionStage<Void> delayStart() {
```
While this is okay... I am not liking where this code is residing. This is putting too much logic in the component registry itself. The registry should be more about registering components, not knowing inner details of them.

I wonder is it possible to move this logic to an appropriate interceptor or component? For example we can add the clear logic to the persistence manager and stable topology directly to the LocalTopologyManager/StateTransferManager. Then we would need to disconnect start method invocation from the actual interceptor initialization and leave the state as INITIALIZING until it is done. I _think_ this would keep the logic more closely aligned with where it should be.

wburns reviewed (commented):
@jabolina let me know what you think about these suggestions.

jabolina commented:
@wburns I'll apply the suggestions. The only thing that worries me is if we spread the restoration requirements to different places. Right now, everything is under the component registry, although I agree that this is overloading it too much. I'll try one approach of using a new component only for that. If that is not right, I'll go by putting the logic on each individual component.

jabolina commented:
Creating a new component dedicated only to restoration wasn't nice, so I added the logic to the LocalTopologyManager/LocalCacheStatus. For the change for persistence, I did a small change only and could keep the purge in the PersistenceManager itself, but it is still calling the clear method after the topology is restored. That is, constructing the whole store and then purging it. I'll look into that now while CI runs.

infinispanrelease commented:
Image pushed for Jenkins build [#20](https://ci.infinispan.org/job/Infinispan/job/PR-10479/20/):
```
quay.io/infinispan-test/server:PR-10479
```

jabolina commented:
Should be OK now. Fixed by allowing to stop a registry in the `INITIALIZING` state and started updating the member list in the `CacheStatus#doLeave` even if the current topology is null. Let's see how it goes in the CI.

infinispanrelease commented:
Image pushed for Jenkins build [#21](https://ci.infinispan.org/job/Infinispan/job/PR-10479/21/):
```
quay.io/infinispan-test/server:PR-10479
```

ryanemerson commented:
705 test failures :face_with_thermometer:

jabolina commented:
Pushed a fix, let's see. The failures were all due to blocking calls. I changed the call from `isRunning` to `cacheExists`.

infinispanrelease commented:
Image pushed for Jenkins build [#22](https://ci.infinispan.org/job/Infinispan/job/PR-10479/22/):
```
quay.io/infinispan-test/server:PR-10479
```

infinispanrelease commented:
Image pushed for Jenkins build [#23](https://ci.infinispan.org/job/Infinispan/job/PR-10479/23/):
```
quay.io/infinispan-test/server:PR-10479
```

wburns commented:
> Pushed a fix, let's see. The failures were all due to blocking calls. I changed the call from `isRunning` to `cacheExists`.

Can you expand upon this? I don't think we can just change the method we are testing against here. If anything we may want do this to both methods no?

wburns commented on core/src/main/java/org/infinispan/manager/DefaultCacheManager.java:
```diff
@@ -560,6 +560,9 @@ private <K, V> Cache<K, V> internalGetCache(String cacheName) {
 
    @Override
    public boolean cacheExists(String cacheName) {
+      if (getCacheBlockingCheck != null) {
```
This has to be on the isRunning method. Existance is not the same thing as the cache running. Without a isRunning check to verify the cache is actually running, the getCache method may block unexpectedly. That is the point of this method for tests.

jabolina replied:
The blocking thread I mentioned before was because of this. The `DefaultCacheManager` calls `BlockHoundUtil.pretendBlock()`. I updated it to have the same execution as before but added a topology check. With this check earlier, we don't have to initiate the operation to find the missing members then.

wburns replied:
Looks like this wasn't changed?

jabolina replied:
Sorry, I missed a commit. Should be correct now.

wburns commented on core/src/test/java/org/infinispan/globalstate/ThreeNodeGlobalStatePartialRestartTest.java:
```diff
@@ -0,0 +1,269 @@
+      applyCacheManagerClusteringConfiguration(config);
+
+      StoreConfigurationBuilder scb;
+      if (sifs) {
```
The test shouldn't care if it is sifs or single file store. Please just use SIFS.

wburns commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java:
```diff
@@ -82,7 +83,8 @@ void writeException(HotRodHeader header, Throwable cause) {
          }
       } else if (header != null) {
-         log.exceptionReported(cause);
+         if (cause instanceof MissingMembersException) log.warn(cause.getMessage());
```
I feel like logging that a member is missing as WARN is a bit too high level. This is a valid state if another server is still starting back up.  I would suggest lowering the severity of logging level for this.

wburns replied:
Also side note but we can't use an unknown cause message for WARN as it needs to be i18n supported.

jabolina replied:
I added an additional check here for tracing, I think using ERROR would pollute the log too much.

wburns commented on server/rest/src/main/java/org/infinispan/rest/RestRequestHandler.java:
```diff
@@ -83,8 +84,9 @@ void handleError(ChannelHandlerContext ctx, FullHttpRequest request, Throwable t
-      } else if (cause instanceof CacheConfigurationException || cause instanceof IllegalArgumentException || cause instanceof EncodingException || cause instanceof Json.MalformedJsonException) {
-         if (getLogger().isTraceEnabled()) getLogger().tracef("Request failed: %s", cause);
+      } else if (cause instanceof CacheConfigurationException || cause instanceof IllegalArgumentException || cause instanceof EncodingException || cause instanceof Json.MalformedJsonException || cause instanceof MissingMembersException) {
+         if (cause instanceof MissingMembersException) getLogger().warn(cause.getMessage());
```
Same here, warn is a bit too much.

wburns commented on server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java:
```diff
@@ -116,7 +116,7 @@ public boolean cacheExists(String name) {
 
    private void checkCacheAvailable(String cacheName) {
-      if (!instance.isRunning(cacheName)) {
+      if (!instance.cacheExists(cacheName)) {
```
This has to be isRunning to ensure the getCache method doesn't block. Just because it exists doesn't mean it won't block. I think there is something else happening with the blocking failure.

jabolina replied:
Updated.

wburns commented on server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:
```diff
@@ -759,7 +759,7 @@ private CompletionStage<RestResponse> setCacheAvailability(RestRequest request)
       String availability = request.getParameter("availability");
       // Use EmbeddedCacheManager directly to allow internal caches to be updated
-      if (!invocationHelper.getRestCacheManager().getInstance().isRunning(cacheName))
+      if (!invocationHelper.getRestCacheManager().getInstance().cacheExists(cacheName))
```
Same thing, this needs to be isRunning to ensure getCache doesn't block.

wburns reviewed (changes requested):
I think we may need to discuss the blocking issue you ran into before. The getCache method on non blocking thread MUST always be proceeded by a isRunning check to ensure it won't block. We cannot change this behavior.

jabolina commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -201,11 +201,9 @@ private void blockingStart() {
                rebuildIndex();
                processFreeEntries();
-            } else {
-               clear();
```
Updated.

jabolina replied:
Created https://issues.redhat.com/browse/ISPN-14582

wburns commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/logging/Log.java:
```diff
@@ -116,4 +117,8 @@ public interface Log extends BasicLogger {
    @Message(value = "Client %s keeps providing outdated topology %s", id = 28029)
    void clientNotUpdatingTopology(SocketAddress socketAddress, int topologyId);
+
+   @LogMessage(level = TRACE)
```
We don't need this for TRACE level, only INFO and higher are internationalized.

jabolina replied:
Updated.

wburns commented on server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java:
```diff
@@ -117,6 +119,10 @@ public boolean cacheExists(String name) {
       if (!instance.isRunning(cacheName)) {
+         GlobalComponentRegistry gcr = SecurityActions.getGlobalComponentRegistry(instance);
+         LocalTopologyManager ltm = gcr.getLocalTopologyManager();
+         if (ltm != null) ltm.assertTopologyStable(cacheName);
```
Much better 👍

wburns commented on server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:
```diff
@@ -752,7 +752,7 @@ private CompletionStage<RestResponse> getCacheConfig(RestRequest request) {
       String cacheName = request.variables().get("cacheName");
       // Use EmbeddedCacheManager directly to allow internal caches to be updated
-      if (!invocationHelper.getRestCacheManager().getInstance().isRunning(cacheName))
+      if (!invocationHelper.getRestCacheManager().getInstance().cacheExists(cacheName))
```
This and below can't be like this. The isRunning call is needed to protect the getCache method.

jabolina replied:
Sorry, I missed this one :/ Updated.

wburns commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:
```diff
@@ -987,7 +987,7 @@ public CompletionStage<Long> sizePublisher(IntSet segments, InvocationContext ct
       private <E> Flowable<E> getValuesFlowable(BiFunction<InnerPublisherSubscription.InnerPublisherSubscriptionBuilder<K, I, R>, Map.Entry<Address, IntSet>, Publisher<E>> subToFlowableFunction) {
          return Flowable.defer(() -> {
-            if (!componentRegistry.getStatus().allowInvocations()) {
+            if (!componentRegistry.getStatus().allowInvocations() && !componentRegistry.getStatus().startingUp()) {
```
So I assume the topology exception is thrown later in this case?

jabolina replied:
Yep, once it reaches the `InvocationContextInterceptor`. I am unsure if this is one of the places we could fail earlier and avoid creating/sending all requests.

wburns commented on core/src/main/java/org/infinispan/util/logging/Log.java:
```diff
@@ -2367,4 +2368,10 @@ CacheConfigurationException storeConfiguredHasBothReadAndWriteOnly(String storeC
+   @Message(value = "Recovering cache '%s' but there are missing members, known members %s of a total of %s", id = 689)
+   void recoverFromStateMissingMembers(String cacheName, List<Address> members, int total);
+
+   MissingMembersException recoverFromStateMissingMembers(String cacheName, List<Address> members, String total);
```
This needs a @Message on it afaik.

jabolina replied:
This is a bit of hack, but I can't find it in the docs now :(
Since this has the same signature as the method above, it will inherit the same message and code. I did this so that the warning and exception have the same code. It is hacky because in one, the `total` is int, and in another is a String.

wburns commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java:
```diff
@@ -72,17 +73,22 @@ void writeException(HotRodHeader header, Throwable cause) {
-         log.exceptionReported(cause);
+      } else if (cause instanceof IllegalStateException || isExceptionTrace(cause)) {
+         if (isExceptionTrace(cause)) {
+            if (log.isTraceEnabled()) log.tracef(cause, "Exception reported");
```
Nitpick: Don't need the traceEnabled call and no need for tracef.

wburns commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/logging/Log.java:
```diff
@@ -116,4 +116,5 @@ public interface Log extends BasicLogger {
    @LogMessage(level = WARN)
    @Message(value = "Client %s keeps providing outdated topology %s", id = 28029)
    void clientNotUpdatingTopology(SocketAddress socketAddress, int topologyId);
+
```
Unneeded

wburns reviewed (commented):
Just minor things left, otherwise LGTM.

jabolina commented:
Added another commit for https://issues.redhat.com/browse/ISPN-14414. With this, we can return the caches on state `INITIALIZING` to the console.

wburns commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -345,7 +345,18 @@ private CacheInfo getCacheInfo(RestRequest request, EmbeddedCacheManager cacheMa
             cacheInfo.status = cache.getStatus().toString();
          } else {
-            cacheInfo.status = ComponentStatus.FAILED.toString();
+            switch (cacheHealth) {
```
This isn't quite right, the if statement above prevents this from ever being anything other than FAILED afai can see.

jabolina replied:
Once again, back copy of code :/ Should be ok now. Thanks.

wburns commented:
Changes look fine, just waiting to make sure CI didn't change.

wburns commented:
Integrated into main, thanks @jabolina !
